### PR TITLE
Ana/add alternative fee logic for emoney

### LINKED ISSUE #4050: Add fee denom selector on feesStep in ActionModal
Many users have requested this and I think it is also something our competitors allow: we should enable a fee selector so users can set their fee.

Some challenges:
- Kava and e-Money don't allow discounts: so going any lower than gasEstimate * gasPrice always results in an error and inability to broadcast.

- Some transactions will result in out of gas errors and users will only notice when they cannot find that transaction. Maybe we should capture this errors in the polling logic of our API and display some message in FE. But probably we'd need notifications working to do this properly

### PR DESCRIPTION
**Pull request by Bitcoinera**
Closes #4050

**Description:**

Tested and everything working great. Good idea @faboweb ! Makes a lot of sense

[This ](https://e-money.network/transactions/C32B25A53586FD00E0FA537B6F8A0EC9A2B2BA2BD182C4E6C4C187E5FB66FD9C)is the "mixed" transaction (other token paying the fees)

The ActionModal on feeStep handling a mixed transaction:

Thank you! 🚀

---

For contributor:

- [ ] Added changes entries. Run `yarn changelog` for a guided process.
- [ ] Reviewed `Files changed` in the github PR explorer
- [ ] Attach screenshots of the UI components on the PR description (if applicable)
- [ ] Scope of work approved for big PRs

For reviewer:

- [ ] Manually tested the changes on the UI

### CODE REVIEW

**Comment by Bitcoinera** [comment]
For Kava it doesn't make sense as they have only one token (for now)

**Comment by Bitcoinera** [comment]
I'm thinking, makes actually sense to do the same for Terra. I tried sending a transaction with "SetMax" and it failed:

`	
TX failed: 23974A35D757000D7460989EE167A049048B217285BCDD644D7762FBEB86A71E Error: out of gas in location: WritePerByte; gasWanted: 43000, gasUsed: 44219`

**Comment by faboweb** [comment]
> TX failed: 23974A35D757000D7460989EE167A049048B217285BCDD644D7762FBEB86A71E Error: out of gas in location: WritePerByte; gasWanted: 43000, gasUsed: 44219

This is out of gas not low fees. Need to increase gasEstimate for that TX

File: src/ActionModal/components/ActionModal.vue [hunk 1]
```diff
@@ -85,8 +85,9 @@
           </TmFormGroup>
           <TableInvoice
             :amount="Number(subTotal)"
-            :estimated-fee="estimatedFee"
+            :estimated-fee="alternativeEstimatedFee || estimatedFee"
             :bond-denom="getDenom"
+            :fee-denom="alternativeFeeDenom || getDenom"
```
Comment (faboweb) [inline src/ActionModal/components/ActionModal.vue:90]: instead make fee-denom optional?

File: src/ActionModal/components/ActionModal.vue [hunk 2]
```diff
@@ -85,8 +85,9 @@
           </TmFormGroup>
           <TableInvoice
             :amount="Number(subTotal)"
-            :estimated-fee="estimatedFee"
+            :estimated-fee="alternativeEstimatedFee || estimatedFee"
```
Comment (faboweb) [inline src/ActionModal/components/ActionModal.vue:88]: reduce to one variable? `fee`?

File: src/ActionModal/components/ActionModal.vue [hunk 3]
```diff
@@ -830,6 +844,20 @@ export default {
     maxDecimals(value, decimals) {
       return Number(BigNumber(value).toFixed(decimals)) // TODO only use bignumber
     },
+    alternativeFeeSelector() {
+      const alternativeFeeBalance = this.balances.find(
+        ({ denom }) =>
+          denom !== this.getDenom && this.amount >= this.estimatedFee
```
Comment (faboweb) [inline src/ActionModal/components/ActionModal.vue:850]: the fee is different in every denom I gues

File: src/ActionModal/components/ActionModal.vue [hunk 4]
```diff
@@ -830,6 +844,20 @@ export default {
     maxDecimals(value, decimals) {
       return Number(BigNumber(value).toFixed(decimals)) // TODO only use bignumber
     },
+    alternativeFeeSelector() {
```
Comment (faboweb) [inline src/ActionModal/components/ActionModal.vue:847]: make this a computed

File: src/ActionModal/components/ActionModal.vue (hunk 3 above)
Comment (Bitcoinera) [reply src/ActionModal/components/ActionModal.vue:850]: oh good point!

**Comment by Bitcoinera** [comment]
> > TX failed: 23974A35D757000D7460989EE167A049048B217285BCDD644D7762FBEB86A71E Error: out of gas in location: WritePerByte; gasWanted: 43000, gasUsed: 44219
> 
> This is out of gas not low fees. Need to increase gasEstimate for that TX

true!

**Comment by faboweb** [comment]
Total should display like this:

 Total  ...  eCHF
            ...  eNOK

**Comment by faboweb** [comment]
This logic could also move to the api

**Comment by Bitcoinera** [comment]
> Total should display like this:
> 
> Total ... eCHF
> ... eNOK

What do you mean?

File: src/ActionModal/components/ActionModal.vue (hunk 1 above)
Comment (Bitcoinera) [reply src/ActionModal/components/ActionModal.vue:90]: actually I just noticed this wasn't needed, as it was already optional

File: src/ActionModal/components/ActionModal.vue (hunk 2 above)
Comment (Bitcoinera) [reply src/ActionModal/components/ActionModal.vue:88]: yes, let's do that

File: src/ActionModal/components/ActionModal.vue (hunk 4 above)
Comment (Bitcoinera) [reply src/ActionModal/components/ActionModal.vue:847]: I cannot since I overwrite `this.gasPrice` and `this.gasEstimate` in there

**Comment by Bitcoinera** [comment]
> This logic could also move to the api

Could be, but then we need to create the API PR first. It is not costly at all, so I think it is fine to have it in here to start with

**Comment by faboweb** [comment]
> This logic could also move to the api

Sorry I will try to flag out of scope ideas better in the future, sorry

**Comment by faboweb** [comment]
breaks e2e tests

**Comment by faboweb** [comment]
User is somehow not signed in, in this PR in the E2E tests

**Comment by Bitcoinera** [comment]
Let's merge 🤩 (in the second to last commit E2E tests did pass)

File: src/ActionModal/components/ActionModal.vue (hunk 4 above)
Comment (faboweb) [reply src/ActionModal/components/ActionModal.vue:838]: which is a bad sign

File: src/ActionModal/components/TableInvoice.vue [hunk 5]
```diff
@@ -1,20 +1,25 @@
...
-          {{ bondDenom }}
+          {{ feeDenom || bondDenom }}
         </span>
       </li>
       <li class="total-row">
         <span>Total</span>
-        <span> {{ total | fullDecimals }} {{ bondDenom }} </span>
+        <div class="total-column">
+          <p>{{ total | fullDecimals }} {{ bondDenom }}</p>
+          <p v-if="feeDenom">
```
Comment (faboweb) [inline src/ActionModal/components/TableInvoice.vue:19]:
```suggestion
          <p v-if="feeDenom !== bondDenom">
```

File: src/ActionModal/components/TableInvoice.vue [hunk 6]
```diff
@@ -49,7 +58,9 @@ export default {
       return this.amount
     },
     total() {
-      return this.estimatedFee + this.subTotal
+      // if there is a feeDenom, it means that subTotal and estimatedFee are different currencies and
+      // cannot be therefore added up together
+      return this.feeDenom ? this.subTotal : this.estimatedFee + this.subTotal
```
Comment (faboweb) [inline src/ActionModal/components/TableInvoice.vue:63]:
```suggestion
      return this.feeDenom !== this.bondDenom ? this.subTotal : this.estimatedFee + this.subTotal
```

**Comment by faboweb** [comment]
Small adjustments to do, then let's merge. I think we should deprecate all the logic and put it into the API.

File: src/ActionModal/components/TableInvoice.vue (hunk 5 above)
Comment (Bitcoinera) [reply src/ActionModal/components/TableInvoice.vue:19]: Shouldn't it rather be `feeDenom && feeDenom !== bondDenom`?

**Comment by Bitcoinera** [comment]
> Small adjustments to do, then let's merge. I think we should deprecate all the logic and put it into the API.

I will open an issue once this is merged, don't worry. It is happening

File: src/ActionModal/components/TableInvoice.vue (hunk 5 above)
Comment (faboweb) [reply src/ActionModal/components/TableInvoice.vue]: feeDenom defaults to `""`

Comment (faboweb) [reply src/ActionModal/components/TableInvoice.vue]: yes your right

**Comment by faboweb** [comment]
In theory this code needs to be tested. But as we move this to the API I will ignore that.

**Comment by Bitcoinera** [comment]
> In theory this code needs to be tested. But as we move this to the API I will ignore that.

I tested it all the way 😉 (like I nearly always do)

**Comment by faboweb** [comment]
I meant unit tests

**Comment by Bitcoinera** [comment]
> I meant unit tests

oh true. But yes, not needed since we are moving it right away 👍

**Pull request merged**